test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Build.AspNet/Formatter/ODataEntityReferenceLinkE2ETests.cs

```diff
@@ -0,0 +1,276 @@
+
+            request.Content = new StringContent(Payload);
+            request.Content.Headers.ContentType = MediaTypeWithQualityHeaderValue.Parse("application/json");
+
+            // Act
+            HttpResponseMessage response = await Client.SendAsync(request);
+
+            // Assert
+            Assert.True(response.IsSuccessStatusCode);
+            Assert.Equal(HttpStatusCode.Created, response.StatusCode);
+            Assert.NotNull(response.Content);
+
+            //Get the above saved entity from the database
+            //and expand its navigation property to see if it was created with
+            //the existing entity correctly.
+            //Also note that we were able to update the name property 
+            //of the existing entity
+            string query = string.Format("{0}/odata/Books?$expand=Author", BaseAddress);
+            HttpRequestMessage requestMessage = new HttpRequestMessage(HttpMethod.Get, query);
+            request.Headers.Accept.Add(MediaTypeWithQualityHeaderValue.Parse("application/json"));
+
+            // Act
+            HttpResponseMessage res = await Client.SendAsync(requestMessage);
+
+            // Assert
+            Assert.True(res.IsSuccessStatusCode);
+            var responseObject = JObject.Parse(await res.Content.ReadAsStringAsync());
+            var result = responseObject["value"] as JArray;
+            var expandProp = result[0]["Author"] as JObject;
+            Assert.Equal(1, expandProp["Id"]);
+            Assert.Equal("UpdatedAuthor", expandProp["Name"]);
+        }
+
+        private async Task ResetDataSource()
+        {
+            string requestUri = BaseAddress + "/odata/ResetDataSource";
+            HttpClient client = new HttpClient();
+            HttpResponseMessage response = await client.GetAsync(requestUri);
+            response.EnsureSuccessStatusCode();
+        }
+    }
+
+    public class BooksController : TestODataController
+#if NETCORE
+        , IDisposable
+#endif
+    { 
+        private ODataEntityReferenceLinkContext db = new ODataEntityReferenceLinkContext();
+
+        [EnableQuery]
+        public ITestActionResult Get()
+        {        
+            return Ok(db.Books);
+        }
+        public ITestActionResult Post([FromBody] Book book)
+        {
+            if (!ModelState.IsValid)
+            {
+                return BadRequest();
+            }
+
+            db.Authors.Attach(book.Author);
+            db.Books.Add(book);
+            db.SaveChanges();
+
+            return Created(book);
+        }
+
+        [HttpPost]
+        [ODataRoute("RelateToExistingEntityAndUpdate")]
+        public ITestActionResult RelateToExistingEntityAndUpdate(ODataActionParameters odataActionParameters)
+        {
+            if (!ModelState.IsValid)
+            {
+                return BadRequest();
+            }
+
+            Book book = (Book)odataActionParameters["book"];
```

Comment:
```suggestion
            Book book = odataActionParameters["book"] as Book; 
```
then check for nulls, 

Interested in how nulls would be treated in this pr. or if someone sent a wrong payload.

Reply:
I could have missed it during an earlier review.